pkg/agent/manager/manager_test.go

```diff
@@ -1,160 +1,883 @@
+	}
+	mgr, err := New(c)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	m := mgr.(*manager)
+	err = m.Start()
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer m.Shutdown()
+
+	cert, key := m.getBaseSVIDEntry()
+	if !cert.Equal(baseSVID) {
+		t.Error("SVID is not equals to configured one")
+	}
+	if key != baseSVIDKey {
+		t.Error("PrivateKey is not equals to configured one")
+	}
+
+	me := m.MatchingEntries(cache.Selectors{&common.Selector{Type: "unix", Value: "uid:1111"}})
+	if len(me) != 2 {
+		t.Error("expected 2 entries")
+	}
+
+	err = compareRegistrationEntries(
+		regEntriesMap["resp2"],
+		[]*common.RegistrationEntry{me[0].RegistrationEntry, me[1].RegistrationEntry})
+	if err != nil {
+		t.Error(err)
+	}
+
+	util.RunWithTimeout(t, 1*time.Second, func() {
+		done := make(chan struct{})
+		wu := m.Subscribe(cache.Selectors{&common.Selector{Type: "unix", Value: "uid:1111"}}, done)
+		u := <-wu
+
+		if len(u.Entries) != 2 {
+			t.Error("expected 2 entries")
+		}
+
+		if len(u.Bundle) != 1 {
+			t.Error("expected 1 bundle")
+		}
+
+		if !u.Bundle[0].Equal(apiHandler.bundle[0]) {
+			t.Error("received bundle should be equals to the server bundle")
+		}
+
+		err := compareRegistrationEntries(
+			regEntriesMap["resp2"],
+			[]*common.RegistrationEntry{u.Entries[0].RegistrationEntry, u.Entries[1].RegistrationEntry})
+		if err != nil {
+			t.Error(err)
+		}
+	})
+}
+
+func TestSVIDRotation(t *testing.T) {
+	dir := createTempDir(t)
+	defer removeTempDir(dir)
+
+	trustDomain := "example.org"
+
+	apiHandler := newMockNodeAPIHandler(&mockNodeAPIHandlerConfig{
+		t:                 t,
+		trustDomain:       trustDomain,
+		dir:               dir,
+		fetchSVIDResponse: fetchSVIDResponseForTestSVIDRotation,
+		svidTTL:           3,
+	})
+	apiHandler.start()
+	defer apiHandler.stop()
+
+	baseTTL := 3 * time.Second
+	baseSVID, baseSVIDKey := apiHandler.newSVID("spiffe://"+trustDomain+"/spire/agent/join_token/abcd", baseTTL)
+
+	c := &Config{
+		ServerAddr: &net.UnixAddr{
+			Net:  "unix",
+			Name: apiHandler.sockPath,
+		},
+		SVID:          baseSVID,
+		SVIDKey:       baseSVIDKey,
+		Log:           testLogger,
+		TrustDomain:   url.URL{Host: trustDomain},
+		SVIDCachePath: path.Join(dir, "svid.der"),
+		Bundle:        apiHandler.bundle,
+		Tel:           &telemetry.Blackhole{},
+	}
+	mgr, err := New(c)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	m := mgr.(*manager)
+	m.rotationFreq = baseTTL / 2
+	m.syncFreq = 1 * time.Hour
+	err = m.Start()
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer m.Shutdown()
+
+	cert, key := m.getBaseSVIDEntry()
+	if !cert.Equal(baseSVID) {
+		t.Error("SVID is not equals to configured one")
+		return
+	}
+	if key != baseSVIDKey {
+		t.Error("PrivateKey is not equals to configured one")
+		return
+	}
+
+	// Loop until we detect a rotation
+	util.RunWithTimeout(t, 2*m.rotationFreq, func() {
+		for cert, _ = m.getBaseSVIDEntry(); cert.Equal(baseSVID); cert, _ = m.getBaseSVIDEntry() {
```

Comment:
I'll try to make this easier to read. The idea was to detect an SVID rotation as soon as possible, so that's why I didn't put any sleep there. Do you think I should put some sleep anyway?